lldb/tools/lldb-dap/lldb-dap.cpp

```diff
@@ -5058,72 +5018,187 @@ int main(int argc, char *argv[]) {
+        auto now = std::chrono::duration<double>(
+            std::chrono::system_clock::now().time_since_epoch());
+        *log << llvm::formatv("{0:f9}", now.count()).str()
+             << " client connected: " << name << "\n";
+      }
+
+      // Move the client into a background thread to unblock accepting the next
+      // client.
+      std::thread client([=, &active_dap_sessions_mutext, &active_dap_sessions,
+                          sock = std::move(sock)]() {
+        llvm::set_thread_name(name + ".runloop");
+        StreamDescriptor input =
+            StreamDescriptor::from_socket(sock->GetNativeSocket(), false);
+        // Close the output last for the best chance at error reporting.
+        StreamDescriptor output =
+            StreamDescriptor::from_socket(sock->GetNativeSocket(), false);
+        DAP dap = DAP(name, program_path, log, std::move(input),
+                      std::move(output), default_repl_mode, pre_init_commands);
+
+        if (auto Err = dap.ConfigureIO()) {
+          llvm::logAllUnhandledErrors(std::move(Err), llvm::errs(),
+                                      "Failed to configure stdout redirect: ");
+          return;
+        }
+
+        RegisterRequestCallbacks(dap);
+
+        {
+          std::scoped_lock lock(active_dap_sessions_mutext);
+          active_dap_sessions.insert(&dap);
+        }
+
+        if (auto Err = dap.Loop()) {
+          llvm::logAllUnhandledErrors(std::move(Err), llvm::errs(),
+                                      "DAP session error: ");
+        }
+
+        {
+          std::scoped_lock lock(active_dap_sessions_mutext);
+          active_dap_sessions.erase(&dap);
+        }
+
+        if (log) {
+          auto now = std::chrono::duration<double>(
+              std::chrono::system_clock::now().time_since_epoch());
+          *log << llvm::formatv("{0:f9}", now.count()).str()
+               << " client closed: " << name << "\n";
+        }
+      });
+      client.detach();
+    });
+    if (auto Err = handle.takeError()) {
+      llvm::logAllUnhandledErrors(std::move(Err), llvm::errs(),
+                                  "Registering accept handler failed: ");
+      return EXIT_FAILURE;
+    }
+
+    error = g_loop.Run();
+    if (error.Fail()) {
+      llvm::logAllUnhandledErrors(error.takeError(), llvm::errs(),
+                                  "MainLoop failed: ");
+      return EXIT_FAILURE;
+    }
+
+    if (log)
+      *log << "lldb-dap server shutdown requested, disconnecting remaining "
+              "clients...\n";
+
+    bool client_failed = false;
+    std::scoped_lock lock(active_dap_sessions_mutext);
+    for (auto *dap : active_dap_sessions) {
+      auto error = dap->Disconnect();
+      if (error.Fail()) {
+        client_failed = true;
+        llvm::errs() << "DAP client " << dap->name
+                     << " disconnected failed: " << error.GetCString() << "\n";
+      }
+    }
+
+    return client_failed ? EXIT_FAILURE : EXIT_SUCCESS;
```

Comment:
There's a race here between this thread exiting (and destroying the `active_dap_sessions` list) and the now-disconnected threads trying to remove themselves from it.

I think that could be fixed by waiting until this list becomes empty (ideally using the aforementioned [std::notify_all_at_thread_exit](https://en.cppreference.com/w/cpp/thread/notify_all_at_thread_exit) to delay cond_var notification until the thread is well into shutting itself down.

Reply:
Done.